File: templates/03_tree_traversal/template.py
"""
Time Complexity：O(N)
Space Complexity：
- DFS: O(H)，H 為樹高
- BFS: O(W)，W 為樹的最大寬度
"""

from typing import List, Optional
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def dfs_preorder(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        def dfs(node):
            if not node:
                return
            res.append(node.val)  # 根
            dfs(node.left)  # 左
            dfs(node.right)  # 右

        dfs(root)
        return res

    def dfs_inorder(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        def dfs(node):
            if not node:
                return
            dfs(node.left)  # 左
            res.append(node.val)  # 根
            dfs(node.right)  # 右

        dfs(root)
        return res

    def dfs_postorder(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        def dfs(node):
            if not node:
                return
            dfs(node.left)  # 左
            dfs(node.right)  # 右
            res.append(node.val)  # 根

        dfs(root)
        return res

    def bfs_level_order(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        res = []
        q = deque([root])

        while q:
            level_size = len(q)  # 1. 鎖定當層的節點數量
            level_vals = []

            for _ in range(level_size):
                node = q.popleft()
                level_vals.append(node.val)

                # 2. 將下一層的子節點排入佇列
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)

            res.append(level_vals)

        return res

File: templates/03_tree_traversal/template.cpp
/*
 * Time Complexity：O(N)
 * Space Complexity：
 * - DFS: O(H)，H 為樹高
 * - BFS: O(W)，W 為樹的最大寬度
 */

#include <queue>
#include <vector>
using namespace std;

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class TreeTraversalTemplates {
 public:
  void dfsPreOrder(TreeNode* root, vector<int>& res) {
    if (root == nullptr) return;

    res.push_back(root->val);       // 先處理自己 (Top-down)
    dfsPreOrder(root->left, res);   // 左
    dfsPreOrder(root->right, res);  // 右
  }

  void dfsInOrder(TreeNode* root, vector<int>& res) {
    if (root == nullptr) return;

    dfsInOrder(root->left, res);   // 左
    res.push_back(root->val);      // 處理自己 (用於 BST 排序)
    dfsInOrder(root->right, res);  // 右
  }

  void dfsPostOrder(TreeNode* root, vector<int>& res) {
    if (root == nullptr) return;

    dfsPostOrder(root->left, res);   // 左
    dfsPostOrder(root->right, res);  // 右
    res.push_back(root->val);        // 最後處理自己 (Bottom-up 收集子樹結果)
  }

  vector<vector<int>> bfsLevelOrder(TreeNode* root) {
    vector<vector<int>> res;
    if (root == nullptr) return res;

    queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
      int size = q.size();  // 1. 鎖定當層的節點數量
      vector<int> level;    // 儲存當層的值

      for (int i = 0; i < size; i++) {
        TreeNode* node = q.front();
        q.pop();

        level.push_back(node->val);

        // 2. 將下一層的子節點排入佇列
        if (node->left != nullptr) q.push(node->left);
        if (node->right != nullptr) q.push(node->right);
      }
      res.push_back(level);
    }
    return res;
  }
};